File: opcua-webservice-spring/examples/uawebservicetest/src/main/java/org/opcfoundation/uawebservicetest/testdigitaltwin/ReferenceElementTestType.java
package org.opcfoundation.uawebservicetest.testdigitaltwin;

import org.eclipse.milo.opcua.stack.core.types.builtin.LocalizedText;
import org.opcfoundation.webserver.digitaltwin.DigitalTwinSpace;
import org.opcfoundation.webserver.digitaltwin.element.ReferenceElementType;
import org.opcfoundation.webserver.service.message.digitaltwin.GetDescriptorRequest;
import org.opcfoundation.webserver.service.message.digitaltwin.GetDescriptorResponse;
import org.opcfoundation.webserver.service.message.digitaltwin.GetLinkRequest;
import org.opcfoundation.webserver.service.message.digitaltwin.GetLinkResponse;
import org.opcfoundation.webserver.types.digitaltwin.ReferenceTargetDescriptor;
import java.util.concurrent.CompletableFuture;

public class ReferenceElementTestType extends ReferenceElementType {

    public ReferenceElementTestType(DigitalTwinSpace space) {
        super("ReferenceElementTestType",new LocalizedText("ReferenceElementTestType"), space);
    }

    @Override
    public CompletableFuture<GetLinkResponse> onGetLinks(GetLinkRequest request)
    {
        GetLinkResponse response = new GetLinkResponse();

        for (int i = request.getOffset(); i<request.getOffset()+5; ++i)
        {
            response.add(new ReferenceTargetDescriptor(
                    Integer.toString(i),
                    new LocalizedText("DT " + i),
                    DigitalTwinSpaceTest.digitalTwinTestType));
        }

        response.setContainsMoreData(0==request.getOffset());
        return CompletableFuture.completedFuture(response);
    }

    @Override
    public CompletableFuture<GetDescriptorResponse> onGetDescriptor(GetDescriptorRequest request) {
        return CompletableFuture.supplyAsync(() -> {
            return new GetDescriptorResponse(
                    new LocalizedText("ReferenceElement-" + request.getId()),
                    new LocalizedText("ReferenceElement with id " + request.getId()));
        });
    }
}

File: opcua-webservice-node/examples/testserver/testdigitaltwin/ReferenceElementTestType.ts
import { UaLocalizedText } from "opcua-webapi-ts";
import { DigitalTwinSpace } from "../../../src/server/digitaltwin";
import { ReferenceElementType } from "../../../src/server/digitaltwin/element";
import { GetDescriptorRequest, GetDescriptorResponse, GetLinkRequest, GetLinkResponse } from "../../../src/server/service/message";
import { ReferenceTargetDescriptor } from "../../../src/server/types/digitaltwin/ReferenceTargetDescriptor";
import { DigitalTwinSpaceTest } from "./DigitalTwinSpaceTest";

export class ReferenceElementTestType extends ReferenceElementType {
    constructor(space: DigitalTwinSpace) {
        super("ReferenceElementTestType", new UaLocalizedText("ReferenceElementTestType"), space);
        this.description = new UaLocalizedText("ReferenceElementTestType");
    }

    override async onGetLinks(request: GetLinkRequest): Promise<GetLinkResponse> {
        const response = new GetLinkResponse();

        for (let index = request.offset; index < request.offset + 5; ++index) {
            response.add(
                new ReferenceTargetDescriptor(
                    `${index}`,
                    new UaLocalizedText(`DT ${index}`),
                    DigitalTwinSpaceTest.digitalTwinTestType,
                ),
            );
        }

        response.containsMoreData = request.offset === 0;
        return response;
    }

    override async onGetDescriptor(request: GetDescriptorRequest): Promise<GetDescriptorResponse> {
        return new GetDescriptorResponse(
            new UaLocalizedText(`ReferenceElement-${request.id}`),
            new UaLocalizedText(`ReferenceElement with id ${request.id}`),
        );
    }
}